### PR TITLE
Np 44638 handler for storing credentials

### PR DESCRIPTION
**Pull request by anetteOlli**

### CODE REVIEW

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java [hunk 1]
```diff
@@ -1,29 +1,86 @@
...
+    private void validateInput(OrcidCredentials input, RequestInfo requestInfo)
+        throws ForbiddenException, BadGatewayException, UnauthorizedException {
+        var userName = requestInfo.getUserName();
+        var orcidFromCristin =
+            attempt(() -> userOrcidResolver.getOrcidForUser(userName)).orElseThrow(
+                fail -> new BadGatewayException("Could not contact cristin API"));
+        if (orcidFromCristin.isEmpty()) {
+            throw new ForbiddenException();
+        }
+        if (!input.getOrcid().toString().contains(orcidFromCristin.get())) {
```
Comment (brinxmat) [inline orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java:75]: I'm going to be a pedant and ask that you write a test where NPE is checked in both input and retrieved cases (the retrieved case is probably covered already)

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java [hunk 2]
```diff
@@ -1,29 +1,86 @@
...
+        this.userOrcidResolver = userOrcidResolver;
     }
 
     @Override
-    protected String processInput(String input, RequestInfo requestInfo, Context context) throws ApiGatewayException {
-        logger.info(input);
-        return input;
+    protected Void processInput(OrcidCredentials input, RequestInfo requestInfo, Context context)
+        throws ApiGatewayException {
+        logger.info(input.getOrcid().toString());
```
Comment (brinxmat) [inline orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java]: Can you add a string that explains what is being logged?

**Comment by anetteOlli** [comment]
Kjetil har poengtert svakheter her allerede i PR'en jeg har på frontenden:
Nå må frontenden gjøre dette i tre trinn:
1. snakke med orcid og sjekke at credentialsene er ekte.
2. Snakke med cristin og lagre orciden på personen
3. Så lagre credentialsene.

Det er altså 3 trinn som kan feile uten at brukeren kan gjøre noe som helst fornuftig med feilene.

Kan vi gjøre noe for å forbedre dette?

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                   .orElseThrow(this::handleFailure);
+    }
+
+    private URI craftUserUri(String userName) throws URISyntaxException {
+        var userId =
+            Arrays.stream(userName.split(USERNAME_DELIMITER))
+                .findFirst().get();
+        return
+            new URIBuilder().setHost(apiHost).setPath("/cristin/person/" + userId).setScheme(HTTPS_SCHEME).build();
+    }
```
Comment (brinxmat) [inline orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: Full-on O-man style:

```
    private URI craftUserUri(String userName) throws URISyntaxException {
        return new URIBuilder().setHost(apiHost)
                .setPath(constructPersonPath(userName))
                .setScheme(HTTPS_SCHEME)
                .build();
    }

    private static String constructPersonPath(String userName) {
        return CRISTIN_PERSON_PATH + extractUserIdentifierFromUserName(userName);
    }

    private static String extractUserIdentifierFromUserName(String userName) {
        return userName.split(USERNAME_DELIMITER)[0];
    }
```

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    private RuntimeException handleFailure(Failure<Optional<String>> fail) {
+        return new CristinPersonApiClientException(fail.getException());
+    }
+
+    private Optional<String> getOrcidFromResponse(HttpResponse<String> response) throws JsonProcessingException {
+        var responseBody = getBodyFromResponse(response);
+        return getOrcidIdentifierFromResponsBody(responseBody);
+    }
+
+    private Optional<String> getOrcidIdentifierFromResponsBody(String responseBody) throws JsonProcessingException {
```
Comment (brinxmat) [inline orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: Typo

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            throw new CristinPersonApiClientException(CRISTIN_API_ERROR_MESSAGE + response.statusCode());
+        }
+        return response.body();
+    }
+
+    private RuntimeException handleFailure(Failure<Optional<String>> fail) {
+        return new CristinPersonApiClientException(fail.getException());
+    }
+
+    private Optional<String> getOrcidFromResponse(HttpResponse<String> response) throws JsonProcessingException {
```
Comment (brinxmat) [inline orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: `get` -> reserved for bean property getters, probably `extract`

Comment (brinxmat) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: Here and elsewhere

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        HttpResponse<String> response;
+        do {
+            if (numberOfAttempts != 0) {
+                Thread.sleep(200);
+            }
+            response = httpClient.send(httpRequest, BodyHandlers.ofString());
+            statusCode = response.statusCode();
+            numberOfAttempts++;
+        } while (numberOfAttempts < MAX_TRIES && statusCode >= HttpURLConnection.HTTP_INTERNAL_ERROR);
+        return response;
```
Comment (brinxmat) [inline orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: Looks like a candidate for resilience Retry

File: orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java [hunk 7]
```diff
@@ -1,29 +1,194 @@
...
+import no.unit.nva.testutils.HandlerRequestBuilder;
+import nva.commons.apigateway.GatewayResponse;
+import nva.commons.core.ioutils.IoUtils;
+import nva.commons.core.paths.UriWrapper;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
-public class StoreOrcidCredentialsFunctionTest {
+@WireMockTest(httpsEnabled = true)
+public class StoreOrcidCredentialsFunctionTest extends OrcidLocalTestDatabase {
```
Comment (brinxmat) [inline orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java]: visibility

Comment (brinxmat) [reply orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java]: Also for functions

File: orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java [hunk 8]
```diff
@@ -1,29 +1,194 @@
...
+
+        var persistedOrcidCredentials = orcidService.fetchOrcidCredentialsByOrcid(orcidCredentials.getOrcid());
+        assertThat(persistedOrcidCredentials.getModified(), is(notNullValue()));
+        assertThat(persistedOrcidCredentials.getCreated(), is(notNullValue()));
+        assertThat(persistedOrcidCredentials.getCreated(), is(equalTo(persistedOrcidCredentials.getModified())));
+
+        persistedOrcidCredentials.setCreated(null);
+        persistedOrcidCredentials.setModified(null);
+
+        assertThat(persistedOrcidCredentials, is(equalTo(orcidCredentials)));
```
Comment (brinxmat) [inline orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java]: persistedOrcidCredentials.hasSameCredentials

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java (hunk 1 above)
Comment (anetteOlli) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java:75]: Going to argue with your here: I have setup api-gateway rules for this already. 

image

Api-gateway already does the null check for me.

Comment (anetteOlli) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java:75]: You get the same error if you try without orcid in the request body.

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java (hunk 2 above)
Comment (anetteOlli) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java]: fixed

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java (hunk 3 above)
Comment (anetteOlli) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: fixed

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java (hunk 4 above)
Comment (anetteOlli) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: fixed

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java (hunk 5 above)
Comment (anetteOlli) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: fixed

File: orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java (hunk 7 above)
Comment (anetteOlli) [reply orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java]: fixed

File: orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java (hunk 8 above)
Comment (anetteOlli) [reply orcid-rest/src/test/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunctionTest.java]: fixed

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java (hunk 6 above)
Comment (anetteOlli) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java]: I agree.

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    private String extractCristinResponseWithRetries(HttpRequest httpRequest) {
+        var retryRegistry = RetryRegistry.ofDefaults();
+        var retry = retryRegistry.retry("get cristin person");
+        Supplier<String> supplier = () -> sendGetCristinPersonRequest(httpRequest);
+        return Try.ofSupplier(Retry.decorateSupplier(retry, supplier)).get();
+    }
+
+    private String sendGetCristinPersonRequest(HttpRequest httpRequest) {
```
Comment (anetteOlli) [inline orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java:92]: I think it's a bit weird mixing the vavr Try methods with the NVA attempts.

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java (hunk 1 above)
Comment (brinxmat) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/StoreOrcidCredentialsFunction.java:75]: I guess the operative word in what I wrote was _pedant_ :D

File: orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java (hunk 9 above)
Comment (brinxmat) [reply orcid-rest/src/main/java/no/sikt/nva/orcid/handlers/UserOrcidResolver.java:92]: It is, we should remove attempt, I guess.

**Review by brinxmat — APPROVED** [review]

**Review by truhacevkir — APPROVED** [review]

**Pull request merged**